mailpoet/lib/Statistics/Track/SubscriberActivityTracker.php

```diff
@@ -58,23 +61,41 @@ private function processTracking(SubscriberEntity $subscriber): void {
   }
 
   private function shouldTrack() {
-    if (!$this->trackingConfig->isCookieTrackingEnabled()) {
+    $timestamp = $this->getLatestTimestamp();
+    // Cookie tracking is disabled and there is no logged-in subscriber who could be used to determine last activity timestamp
+    if ($timestamp === null) {
       return false;
     }
-    $timestamp = $this->pageViewCookie->getPageViewTimestamp() ?? 0;
     return $timestamp + self::TRACK_INTERVAL < $this->wp->currentTime('timestamp');
   }
 
+  private function getLatestTimestamp(): ?int {
+    if ($this->trackingConfig->isCookieTrackingEnabled()) {
+      return $this->pageViewCookie->getPageViewTimestamp() ?? 0;
+    }
+    // In case the cookie tracking is disabled fallback to last engagement of currently logged subscriber
+    $subscriber = $this->getSubscriber();
+    if (!$subscriber) {
+      return null;
+    }
+    return $subscriber->getLastEngagementAt() ? $subscriber->getLastEngagementAt()->getTimestamp() : 0;
+  }
+
   private function getSubscriber(): ?SubscriberEntity {
+    if ($this->activeSubscriber instanceof SubscriberEntity) {
+      return $this->activeSubscriber;
+    }
     $wpUser = $this->wp->wpGetCurrentUser();
     if ($wpUser->exists()) {
-      return $this->subscribersRepository->findOneBy(['wpUserId' => $wpUser->ID]);
+      $this->activeSubscriber = $this->subscribersRepository->findOneBy(['wpUserId' => $wpUser->ID]);
+      return $this->activeSubscriber;
```

Comment:
I'm not sure if this type of caching (`$this->activeSubscriber`) is needed. Could we just call this function once and pass subscriber as an argument? You would then need to put more logic into `trackActivity()` itself, I guess... Not important though.

Reply:
I added it as an optimization for the case when cookie tracking is not allowed and I need to use subscriber's last_engagement_at as a fallback value. The call of `$this->subscribersRepository->findOneBy(['wpUserId' => $wpUser->ID]);` always triggers an SQL query and because of how I structured the class it needs to call it twice. I guess I can refactor it so that the caching is not needed.